pkg/apis/config/v1alpha4/types.go

```diff
@@ -31,6 +31,12 @@ type Cluster struct {
 	// Networking contains cluster wide network settings
 	Networking Networking `yaml:"networking,omitempty"`
 
+	// FeatureGates contains a map of Kubernetes feature gates to wether they
+	// are enabled
+	//
+	// https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/
+	FeatureGates map[string]bool `yaml:"featureGates,omitempty"`
```

Comment:
there is a common separation between k8s feature gates and kubeadm feature gates.
usually the kubeadm feature gates can hold features only relevant to kubeadm.
also they are not the preferred mechanic for feature graduation on our side.

would it make sense to also support kubeadm feature gates within the same iteration?

```
kubernetesFeatureGates
kubeadmFeatureGates
```

Reply:
right now I'm focused on kubernetes component feature gates (motivation is testing with AllBeta=false AllAlpha=false ...), currently it looks like the only kubeadm feature gate documented is related to ipv6 dualstack, which kind will probably choose to toggle based on networking.ipFamily ?

Reply:
yes, kubeadm does not have any interesting feature gates to kind currently.
the alternative for potential future support is:
```
featureGates
kubeadmFeatureGates
```

Reply:
this outmodes the kludge at https://kind.sigs.k8s.io/docs/user/quick-start/#enable-feature-gates-in-your-cluster

Reply:
well , the DualStack job fails because of this
https://github.com/kubernetes-sigs/kind/pull/1255
kubaedm DualStack feature gate pass the feature-gate parameter to all the components, but since this merged this is override because it configures
```
  extraArgs:
    feature-gates: ""
```

based on this I think that I will switch in dual stack to the new config and not patch kubaedm, what do you think?

Reply:
this is how it looks like now the dualstack feature gate
https://github.com/kubernetes-sigs/kind/blob/9405b2570d081f92da9534ca644ae03e4dd876e2/hack/ci/e2e-k8s.sh#L86-L97

Reply:
Yeah, we should only output these here if feature gates is non empty.

Reply:
that was my first thought, but will it merge both feature-gates in case that they are enabled in both places?
I mean, if I configure 
```
 featureGates: 
   EndpointSlice: true 
 kubeadmConfigPatches: 
 - | 
   apiVersion: kubeadm.k8s.io/v1beta2 
   kind: ClusterConfiguration 
   metadata: 
     name: config 
   featureGates: 
     IPv6DualStack: true 
 EOF 
```

will I have endpointslices and dual-stack, or the former will override the later as it is doing now?

Reply:
That's a good question. If kubeadm already has a config option to set kubernetes feature gates uniformly, making kind use that option under the covers (instead of building a feature-gates flag override) would probably make sense. I'm not familiar enough with kubeadm to know what that option is or how it handles merges

Reply:
if the user has:
```
apiVersion: kubeadm.k8s.io/v1beta2
kind: ClusterConfiguration
featureGates: 
     IPv6DualStack: true 
apiServer:
    extraArgs:
        "feature-gates": "EndpointSlice=true"
```

the resulting flag for the kube-apiserver will be `--feature-gates="EndpointSlice=true"`

Reply:
- kubeadm does not have a uniform feature gate option.
- I'm reworking the config generation to not emit the feature gate args in kubeadm config if the kind option is not set.
- using patches for feature-gates is equally as powerful but more verbose, you cannot stack these because it's a string rather than a map (except in component configs)

Reply:
> I'm reworking the config generation to not emit the feature gate args in kubeadm config if the kind option is not set.

That's sensible.

Reply:
> that was my first thought, but will it merge both feature-gates in case that they are enabled in both places?
I mean, if I configure 
```
 featureGates: 
   EndpointSlice: true 
 kubeadmConfigPatches: 
 - | 
   apiVersion: kubeadm.k8s.io/v1beta2 
   kind: ClusterConfiguration 
   metadata: 
     name: config 
   featureGates: 
     IPv6DualStack: true 
 EOF 
```
> will I have endpointslices and dual-stack, or the former will override the later as it is doing now?

So amusingly these are the docs:
> - "featureGates" still exists under ClusterConfiguration, but there are no supported feature gates in 1.13.
> https://godoc.org/k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1beta1

If you read the code though there are at least two feature gates.

If you tell kubeadm to enable `IPv6DualStack` then it will set the Kubernetes components feature gate flags with this **only if the user has not already set the flag** so you need to set the ipv6 feature gate at the kind level if you want both.

E.G. https://github.com/kubernetes/kubernetes/blob/f4b6b751cdf2d46940355359443b263cb883d006/cmd/kubeadm/app/phases/kubelet/flags.go#L124-L128

This is actually easier for kind to handle properly now. What should happen instead is:
```
kind: Cluster
apiVersion: kind.x-k8s.io/v1alpha4
networking:
  ipfamily: ipv4ipv6-dualstack # or whatever we want to call this
```
=> internally default ipv6 featuregate to enabled at the kind config level => passed to all components explicitly.

And users should stop relying on patching the kubeadm configs in N different places to do the equivilant of the new featuregate field.

Reply:
+100